### PR TITLE
Vault secret manager

### PR DESCRIPTION
**Pull request by alvaroaleman**
Adds a vault secret manager, see the provided Readme for a detailled description of how it works. This PR only provides the backend part and no frontend yet. The first commit "Add a vaultclient" is in a dedicated PR: https://github.com/openshift/ci-tools/pull/1737

/cc @openshift/openshift-team-developer-productivity-test-platform 
Ref https://issues.redhat.com/browse/DPTP-1891

### CODE REVIEW

File: cmd/secret-collection-manager/main.go [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+	vaultAddr     string
+	vaultToken    string
+}
+
+func parseOptions() *option {
+	o := &option{}
+	flag.StringVar(&o.kvStorePrefix, "kv-store-prefix", "secret/self-managed", "Vault KV folder under which all policies will get created")
+	flag.StringVar(&o.listenAddr, "listen-addr", "127.0.0.1:8080", "The address to listen on")
+	flag.StringVar(&o.vaultAddr, "vault-addr", "http://127.0.0.1:8300", "The address under which vault should be reached")
+	flag.StringVar(&o.vaultToken, "vault-token", "jpuxZFWWFW7vM882GGX2aWOE", "The privileged token to use when communicating with vault, must be able to CRUD policies")
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:36]: Kind of weird that this has a default

File: cmd/secret-collection-manager/main.go [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+func main() {
+	logrusutil.ComponentInit()
+	o := parseOptions()
+
+	privilegedVaultClient, err := vaultclient.New(o.vaultAddr, o.vaultToken)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to construct vault client")
+	}
+
+	if err := server(privilegedVaultClient, o.kvStorePrefix, o.listenAddr).ListenAndServe(); err != nil && err != http.ErrServerClosed {
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:50]: Might as well wrap with `interrupts.ListenAndServe`?

File: cmd/secret-collection-manager/main.go [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+		logrus.WithError(err).Fatal("failed to listen and serve")
+	}
+
+}
+
+func server(privilegedVaultClient *vaultclient.VaultClient, kvStorePrefix, listenAddr string) *http.Server {
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:56]: What's the use of breaking this func out?

File: cmd/secret-collection-manager/main.go [hunk 4]
```diff
@@ -0,0 +1,415 @@
...
+	if c.ids == nil {
+		c.ids = map[string]string{}
+	}
+	c.names[name] = id
+	c.ids[id] = name
+}
+
+func (m *secretCollectionManager) mux() *httprouter.Router {
+	router := httprouter.New()
+	router.GET("/healthz", healthHandler)
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:127]: We have health & readiness lib already from Prow, prefer that for consistency of code, ports, etc

File: cmd/secret-collection-manager/main.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+		http.Error(w, fmt.Sprintf("failed to create secret collection. RequestID: %s", l.Data["UID"]), 501)
+	}
+}
+
+func (m *secretCollectionManager) createSecretCollection(_ *logrus.Entry, userName, secretCollectionName string) error {
+	user, err := m.userByAliasCached(userName)
+	if err != nil {
+		return fmt.Errorf("failed to get user %s: %w", userName, err)
+	}
+	kvPathSplit := strings.Split(m.kvStorePrefix, "/")
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main.go:229]: Can we unit test string manipulations like this block?

File: cmd/secret-collection-manager/main_test.go [hunk 6]
```diff
@@ -0,0 +1,371 @@
...
+	defer cancel()
+	server := server(client, "secret/self-managed", managerListenAddr)
+	go func() {
+		if err := server.ListenAndServe(); err != nil && ctx.Err() == nil {
+			cancel()
+			t.Errorf("failed to start secret-collection-manager: %v", err)
+		}
+	}()
+
+	if err := waitForServiceReady(ctx, t, fmt.Sprintf("http://%s/healthz", managerListenAddr)); err != nil {
```
Comment (stevekuznetsov) [inline cmd/secret-collection-manager/main_test.go:103]: The e2e framework has code to transparently stop, wait and stop (on timeout) processes that use the k8s ready/health endpoints (so it knows which ports to poke) - might be able to re-use?

File: pkg/vaultclient/vaultclient.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+	}
+	// 404 for list means no results: https://github.com/hashicorp/vault/issues/5861
+	if raw == nil || raw.Data == nil {
+		return nil
+	}
+	return v.dataInto(raw.Data, target)
+}
+
+func (v *VaultClient) readInto(path string, target interface{}) error {
+	raw, err := v.Logical().Read(path)
```
Comment (stevekuznetsov) [inline pkg/vaultclient/vaultclient.go:165]: How are retries, etc handled? Inside of their client?

File: cmd/secret-collection-manager/main.go (hunk 2 above)
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:50]: I would be less hesitant to use `interrupts` if https://github.com/kubernetes/test-infra/issues/17737 was fixed and it wasn't entirely based on package-level variables :/

File: cmd/secret-collection-manager/main.go (hunk 3 above)
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:56]: That func is what the tests use

File: cmd/secret-collection-manager/main.go (hunk 4 above)
Comment (alvaroaleman) [reply cmd/secret-collection-manager/main.go:127]: It doesn't really do anything different other than hardcoding a different port

File: pkg/vaultclient/vaultclient.go (hunk 7 above)
Comment (alvaroaleman) [reply pkg/vaultclient/vaultclient.go]: Yeah, they use go-retriable underneath with a couple of custom settings, I thought it was best to just leave the defaults there

File: cmd/secret-collection-manager/main.go [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+	name := params.ByName("name")
+	if name == "" {
+		http.Error(w, "name url parameter must not be empty", 400)
+		return
+	}
+
+	collections, err := m.getCollectionsForUser(l, user)
+	if err != nil {
+		l.WithError(err).Error("failed to get secret collections for user")
+		http.Error(w, fmt.Sprintf("failed to check if user is allowed to change secret collection. RequestID: %s", l.Data["UID"]), 501)
```
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:159]: Curious, why 501?

File: cmd/secret-collection-manager/main.go [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+			break
+		}
+	}
+	if !found {
+		l.Warn("User tried to update secret collection they don't have access to/that doesn't exist")
+		http.Error(w, fmt.Sprintf("secret collection not found. RequestID: %s", l.Data["UID"]), 404)
+		return
+	}
+
+	var body secretColectionUpdateBody
```
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:176]: typo: `Collection`

File: cmd/secret-collection-manager/main.go [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+	policy := managedVaultPolicy{Path: map[string]managedVaultPolicyCapabiltyList{
+		m.kvMetadataPrefix + "/" + secretCollectionName + "/*": {Capabilities: []string{"list"}},
+		m.kvDataPrefix + "/" + secretCollectionName + "/*":     {Capabilities: []string{"create", "update", "read"}},
+	}}
+	serializedPolicy, err := json.Marshal(policy)
+	if err != nil {
+		return fmt.Errorf("failed to serialize policy for %s: %w", secretCollectionName, err)
+	}
+	if err := m.privilegedVaultClient.Sys().PutPolicy(prefixedName(secretCollectionName), string(serializedPolicy)); err != nil {
+		println(string(serializedPolicy))
```
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:249]: debugging leftover?

File: cmd/secret-collection-manager/main.go [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+	var groupNames []string
+	for _, groupID := range user.GroupIDs {
+		name, err := m.groupNameFromId(groupID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to resolve name for group id %s: %w", groupID, err)
+		}
+		if !strings.HasPrefix(name, objectPrefix) {
+			continue
+		}
+		groupNames = append(groupNames, name)
```
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:313]: I'd consider `if HasPrefix(...) { groupNames = append(...) }` a bit more readable (getting rid of the negation and `continue`) but /shrug

File: cmd/secret-collection-manager/main.go [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+	if n := len(policyData.Path); n != 2 {
+		return nil, fmt.Errorf("policy %s didn't have two but %d paths", group.Policies[0], n)
+	}
+
+	var collection secretCollection
+	for path := range policyData.Path {
+		if !strings.HasPrefix(path, m.kvMetadataPrefix) && !strings.HasPrefix(path, m.kvDataPrefix) {
+			return nil, fmt.Errorf("path %s in policy %s neither had the metadata(%s) nor the data(%s) prefix", path, group.Policies[0], m.kvDataPrefix, m.kvDataPrefix)
+		}
+		collection.Name = strings.TrimSuffix(strings.TrimPrefix(strings.TrimPrefix(path, m.kvMetadataPrefix+"/"), m.kvDataPrefix+"/"), "/*")
```
Comment (petr-muller) [inline cmd/secret-collection-manager/main.go:404]: :exploding_head:

**Comment by alvaroaleman** [comment]
@petr-muller Addressed all your feedback. Also renamed the thing to `vault secret collection manager`, because without the vault I can not remember the name

**Review by petr-muller — APPROVED** [review]
LGTM, holding if you want more reviews, it's a big thing

/hold

**Pull request merged**